### PR TITLE
fix: concurrent modification exception in update method

### LINKED ISSUE #2469: Crash when Characters die from fall damage during server prediction
### What you were trying to do

Move a deer with high movement speed off a cliff

### What actually happened

Server crashes with ConcurrentModificationException, similar to (but distinct from) #2468 since it's a different collection and the event is dispatch from the `update` method of a component system rather than a BT Node.

### How to reproduce

Kill a (non-player) character with fall damage (via CharacterMoveInputEvent).

### Log details and game version
```
"main@1" prio=10 tid=0x1 nid=NA runnable
  java.lang.Thread.State: RUNNABLE
	  at org.terasology.logic.characters.ServerCharacterPredictionSystem.onDestroy(ServerCharacterPredictionSystem.java:97)
	  at org.terasology.logic.characters.ServerCharacterPredictionSystemMethodAccess.invoke(Unknown Source:-1)
	  at org.terasology.entitySystem.event.internal.EventSystemImpl$ByteCodeEventHandlerInfo.invoke(EventSystemImpl.java:506)
	  at org.terasology.entitySystem.event.internal.EventSystemImpl.sendStandardEvent(EventSystemImpl.java:269)
	  at org.terasology.entitySystem.event.internal.EventSystemImpl.send(EventSystemImpl.java:260)
	  at org.terasology.entitySystem.entity.internal.PojoEntityManager.destroy(PojoEntityManager.java:538)
	  at org.terasology.entitySystem.entity.internal.BaseEntityRef.destroy(BaseEntityRef.java:138)
	  at org.terasology.logic.health.EntityDestructionAuthoritySystem.onDestroy(EntityDestructionAuthoritySystem.java:32)
	  at org.terasology.logic.health.EntityDestructionAuthoritySystemMethodAccess.invoke(Unknown Source:-1)
	  at org.terasology.entitySystem.event.internal.EventSystemImpl$ByteCodeEventHandlerInfo.invoke(EventSystemImpl.java:506)
	  at org.terasology.entitySystem.event.internal.EventSystemImpl.sendStandardEvent(EventSystemImpl.java:269)
	  at org.terasology.entitySystem.event.internal.EventSystemImpl.send(EventSystemImpl.java:260)
	  at org.terasology.entitySystem.entity.internal.BaseEntityRef.send(BaseEntityRef.java:145)
	  at org.terasology.logic.health.HealthAuthoritySystem.doDamage(HealthAuthoritySystem.java:185)
	  at org.terasology.logic.health.HealthAuthoritySystem.checkDamage(HealthAuthoritySystem.java:200)
	  at org.terasology.logic.health.HealthAuthoritySystem.onLand(HealthAuthoritySystem.java:241)
	  at org.terasology.logic.health.HealthAuthoritySystemMethodAccess.invoke(Unknown Source:-1)
	  at org.terasology.entitySystem.event.internal.EventSystemImpl$ByteCodeEventHandlerInfo.invoke(EventSystemImpl.java:506)
	  at org.terasology.entitySystem.event.internal.EventSystemImpl.sendStandardEvent(EventSystemImpl.java:269)
	  at org.terasology.entitySystem.event.internal.EventSystemImpl.send(EventSystemImpl.java:260)
	  at org.terasology.entitySystem.entity.internal.BaseEntityRef.send(BaseEntityRef.java:145)
	  at org.terasology.logic.characters.KinematicCharacterMover.walk(KinematicCharacterMover.java:647)
	  at org.terasology.logic.characters.KinematicCharacterMover.updatePosition(KinematicCharacterMover.java:559)
	  at org.terasology.logic.characters.KinematicCharacterMover.step(KinematicCharacterMover.java:103)
	  at org.terasology.logic.characters.ServerCharacterPredictionSystem.stepState(ServerCharacterPredictionSystem.java:168)
	  at org.terasology.logic.characters.ServerCharacterPredictionSystem.onPlayerInput(ServerCharacterPredictionSystem.java:126)
	  at org.terasology.logic.characters.ServerCharacterPredictionSystemMethodAccess.invoke(Unknown Source:-1)
	  at org.terasology.entitySystem.event.internal.EventSystemImpl$ByteCodeEventHandlerInfo.invoke(EventSystemImpl.java:506)
	  at org.terasology.entitySystem.event.internal.EventSystemImpl.sendStandardEvent(EventSystemImpl.java:269)
	  at org.terasology.entitySystem.event.internal.EventSystemImpl.send(EventSystemImpl.java:260)
	  at org.terasology.entitySystem.entity.internal.BaseEntityRef.send(BaseEntityRef.java:145)
	  at org.terasology.flexiblemovement.FlexibleMovementSystem.update(FlexibleMovementSystem.java:37)
	  at org.terasology.engine.modes.StateIngame.update(StateIngame.java:176)
	  at org.terasology.engine.TerasologyEngine.mainLoop(TerasologyEngine.java:413)
	  at org.terasology.engine.TerasologyEngine.run(TerasologyEngine.java:368)
	  - locked <0x1906> (a org.terasology.engine.TerasologyEngine)
	  at org.terasology.engine.Terasology.main(Terasology.java:152)
```

### PR DESCRIPTION
**Pull request by Elyahu41**
### Contains

Fixes #2469

### How to test

Make a deer fall off a cliff

### Outstanding before merging

Should check for null pointers in other areas of the code, I didn't notice any while testing.

### CODE REVIEW

File: engine/src/main/java/org/terasology/logic/characters/ServerCharacterPredictionSystem.java [hunk 1]
```diff
@@ -200,7 +203,8 @@ public void update(float delta) {
                         // Haven't received input in a while, repeat last input
                         CharacterMoveInputEvent lastInput = lastInputEvent.get(entry.getKey());
                         if (lastInput != null) {
-                            CharacterMoveInputEvent newInput = new CharacterMoveInputEvent(lastInput, (int) (time.getGameTimeInMs() - state.getTime()));
+                            CharacterMoveInputEvent newInput = new CharacterMoveInputEvent(lastInput,
+                                    (int) (time.getGameTimeInMs() - state.getTime()));
```
Comment (Elyahu41) [inline engine/src/main/java/org/terasology/logic/characters/ServerCharacterPredictionSystem.java:207]: IntelliJ automatically did this.

File: engine/src/main/java/org/terasology/logic/characters/ServerCharacterPredictionSystem.java [hunk 2]
```diff
@@ -214,11 +218,16 @@ public void update(float delta) {
             if (entry.getKey().equals(localPlayer.getCharacterEntity())) {
                 continue;
             }
-
-            setToTime(renderTime, entry.getKey(), entry.getValue());
```
Comment (Elyahu41) [inline engine/src/main/java/org/terasology/logic/characters/ServerCharacterPredictionSystem.java:218]: the "setToTime" method threw a null pointer exception in the "CharacterMovementSystemUtility" class. So I just skipped the object that was being deleted.

File: engine/src/main/java/org/terasology/logic/characters/ServerCharacterPredictionSystem.java [hunk 3]
```diff
@@ -214,11 +218,16 @@ public void update(float delta) {
...
+            if (!characterStatesToRemove.contains(entry.getKey())) {
+                setToTime(renderTime, entry.getKey(), entry.getValue());
+            }
         }
+        characterStates.keySet().removeAll(characterStatesToRemove);
+        characterStatesToRemove.clear();
```
Comment (Elyahu41) [inline engine/src/main/java/org/terasology/logic/characters/ServerCharacterPredictionSystem.java:226]: forgot to add this in the last PR. My Bad!!!

File: engine/src/main/java/org/terasology/logic/characters/ServerCharacterPredictionSystem.java [hunk 4]
```diff
@@ -16,6 +16,7 @@
 
 package org.terasology.logic.characters;
 
+import com.google.api.client.util.Lists;
```
Comment (DarkWeird) [inline engine/src/main/java/org/terasology/logic/characters/ServerCharacterPredictionSystem.java:19]: Change it on `com.google.common.collect.Lists`

Why is it work for you? It is must be permited by gradle's `implementation`

Comment (Elyahu41) [reply engine/src/main/java/org/terasology/logic/characters/ServerCharacterPredictionSystem.java:19]: Weird, I just let intellij import that class. It wouldn't work without it. I can change it to "com.google.common.collect.Lists" in another PR, if you want.

**Comment by Elyahu41** [comment]
I just created a new PR with the appropriate changes. #4131

**Pull request closed without merging**